Spigot-API-Patches/0167-Add-BaseComponent-array-to-a-variety-of-events.patch

```diff
@@ -0,0 +1,2079 @@
++ * </p>
++ *
++ * <p>
++ * Note that format specifiers are ignored if they are not in any of the following locations:
++ * <ul>
++ * <li>
++ *     {@link TextComponent#getText()}
++ * </li>
++ * <li>
++ *     {@link BaseComponent#getClickEvent()}
++ *     Note that for this the argument is converted to plain text when formatting.
++ * </li>
++ * <li>
++ *     {@link BaseComponent#getHoverEvent()}
++ * </li>
++ * <li>
++ *     {@link BaseComponent#getExtra()}
++ * </li>
++ * </ul>
++ * </p>
++ *
++ * <p>
++ * The locations above are formatted in the order that they are listed.
++ * </p>
++ *
++ * <p>
++ * If a null value is used for any argument, then the output is {@code new TextComponent("null")}
++ * </p>
++ *
++ * <p>
++ * Finally, it is considered invalid if a format specifier is not in one TextComponent instance.
++ * </p>
++ *
++ */
++public final class BaseComponentFormatter {
++
++    /* TODO: Make invalid format exceptions verbose about what is wrong */
++
++    /**
++     * Formats with the specified format according to the specifications laid out by {@link BaseComponentFormatter#isValidFormat(BaseComponent[], int)}.
++     *
++     * <p>
++     *
++     * </p>
++     * @param format The specified format.
++     * @param args The arguments to use.
++     */
++    public static BaseComponent[] format(final BaseComponent[] format, final BaseComponent[]... args) {
++        final List<BaseComponent> ret = new ArrayList<>();
++
++
++        return ret.toArray(new BaseComponent[ret.size()]);
++    }
++
++    private static boolean formatText(final List<BaseComponent> dst, final TextComponent component, final FormatSpecifier specifier, final BaseComponent[]... args) {
++        final String text = component.getText();
++    }
++
++    private static void formatPlainText(final List<BaseComponent> dst, final String text, final FormatSpecifier specifier,
++                                        final BaseComponent previousFormat, final BaseComponent[]... args) {
++        int off, prevOff = 0;
++        for (off = text.indexOf('%'); off != -1;) {
++            // TODO: Test how client handles side-by-side components (does it inherit the formatting?)
++        }
++    }
++
++    /**
++     * Tests if the specified format is valid BaseComponent format.
++     * Note that the entire specification for a BaseComponent format is specified
++     * {@link BaseComponentFormatter here}.
++     *
++     * @param format The specified format to check.
++     * @param arguments The total number of arguments. If any format specifier specifies an argument_index that is
++     *                  greater than this value, the format specifier is considered invalid.
++     * @return true if the specified format contains no invalid format specifiers, false otherwise.
++     */
++    public static boolean isValidFormat(final BaseComponent[] format, final int arguments) {
```

Comment:
be more explicit with what's wrong in the format like String#format

Reply:
Also, minor nit: rename arguments to argumentCount

Reply:
I think it's fine as it is now